samples/Gallery/Pages/TreeView/EditableTreeView.fs

```diff
@@ -0,0 +1,280 @@
+        | AddNodeTo parentNodeName ->
+            let newNode = leaf ""
+
+            let updated =
+                match findNodes (fun n -> n.Name = parentNodeName) model.Nodes |> Seq.tryExactlyOne with
+                | Some node ->
+                    node.Children <- node.Children @ [ newNode ]
+                    model
+                | None ->
+                    { model with
+                        Nodes = model.Nodes @ [ newNode ] }
+
+            updated, Cmd.none
+
+        | RemoveNode node ->
+            { model with
+                Nodes = model.Nodes |> List.choose(removeNode node) },
+            Cmd.none
+
+        | FilterChanged filter -> { model with Filter = filter }, Cmd.none
+
+        | SelectionItemChanged args ->
+            let updated =
+                if args.AddedItems.Count > 0 then
+                    let node = args.AddedItems[0] :?> EditableNode
+                    let modelNode = findNodes (fun n -> n = node) model.Nodes |> Seq.tryExactlyOne
+                    { model with Selected = modelNode }
+                (*  TODO I feel the proper way to handle this event would include the following, but that breaks selection of nodes on the top level.
+                    For some reason a second event removing the selection is triggered for them immediately after selection.
+                    Why is that happening? *)
+                (*else if args.RemovedItems.Count > 0 then
+                    let node = args.RemovedItems[0] :?> EditableNode
+
+                    if model.Selected.IsSome && node = model.Selected.Value then
+                        { model with Selected = None }
+                    else
+                        model*)
+                else
+                    model
+
+            updated, Cmd.none
+
+    let program =
+        Program.statefulWithCmd init update
+        |> Program.withTrace(fun (format, args) -> Debug.WriteLine(format, box args))
+        |> Program.withExceptionHandler(fun ex ->
+#if DEBUG
+            printfn $"Exception: %s{ex.ToString()}"
+            false
+#else
+            true
+#endif
+        )
+
+    /// Functions for leaf nodes for adding a node to a list.
+    module AddLeaf =
+        let private prefix = "addTo!"
+
+        /// Appends an Add leaf Node (referencing the parentNode or the model if None) to the Node list.
+        let appendTo list (parentNode: EditableNode option) =
+            let parentName = parentNode |> Option.map(_.Name) |> Option.defaultValue ""
+            list @ [ leaf(prefix + parentName) ]
+
+        /// Identifies an Add leaf Node.
+        let is (node: EditableNode) = node.Name.StartsWith(prefix)
+
+        /// Retrieves the parent Node Name from the Node Name of an Add leaf Node.
+        let getParentNodeName (node: EditableNode) = node.Name.Substring(prefix.Length)
+
+    let view () =
+        Component(program) {
+            let! model = Mvu.State
+
+            let rec filter (nodes: EditableNode list) =
```

Comment:
I would recommend NOT to add any logic here(Most likely will result in the MVU loop to go out of sync). Better to have proper `Msg` where you can do the filter over the `nodes` and `sub-nodes`.

Reply:
@edgarfgp Not sure I understand how to do that exactly - do you suggest replacing the "functional" filtering with a "statically" filtered copy of the tree in the model that's bound to the TreeView instead?

Reply:
Yes that is what I meant